galley/pkg/config/analysis/analyzers/serviceentry/httpsonhttp.go

```diff
@@ -0,0 +1,64 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceentry
+
+import (
+	"fmt"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+type HTTPSOnHTTPAnalyzer struct{}
+
+var _ analysis.Analyzer = &HTTPSOnHTTPAnalyzer{}
+
+func (serviceEntry *HTTPSOnHTTPAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "serviceentry.HTTPSOnHTTPAnalyzer",
+		Description: "Checks if HTTPS traffic on HTTP port",
+		Inputs: collection.Names{
+			collections.IstioNetworkingV1Alpha3Serviceentries.Name(),
+		},
+	}
+}
+
+func (serviceEntry *HTTPSOnHTTPAnalyzer) Analyze(context analysis.Context) {
+	context.ForEach(collections.IstioNetworkingV1Alpha3Serviceentries.Name(), func(resource *resource.Instance) bool {
+		serviceEntry.analyzeProtocol(resource, context)
+		return true
+	})
+}
+
+func (serviceEntry *HTTPSOnHTTPAnalyzer) analyzeProtocol(resource *resource.Instance, context analysis.Context) {
+	se := resource.Message.(*v1alpha3.ServiceEntry)
+
+	for index, port := range se.Ports {
+		if port.Number == 443 && port.Protocol == "HTTP" {
```

Comment:
Two comments:
* Sending HTTP to 443 may not be a mistake if we are doing TLS origination (https://istio.io/latest/docs/tasks/traffic-management/egress/egress-tls-origination/#tls-origination-for-egress-traffic). Or we have seen a large number of cases of users that have apps listening on plaintext on port 443 (I don't know why...).
* The port that matters here is `targetPort`, we should check that as well. If unset, we can check `port.Number`

Reply:
so when we do TLS origination, we don't see `protocol` in the ServiceEntry section because the port will be upgraded to HTTPS via DestinationRule right (the one care is `port` itself)? I can work to check also if the host is doing TLS origination or not.

a question about `targetPort`. so if we don't set this, the `targetPort` will be the same as the `port` right? so if the serviceentry looks like this one, there is no validation error?

```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: httpbin
spec:
  hosts:
  - httpbin.org
  ports:
  - number: 443
    name: http
    protocol: HTTP
    targetPort: 80
  resolution: DNS
```

Reply:
This is done